File: Data Structures & Algorithms/unboundedKnapsack/submission-2.cpp
class Solution {
public:
    int maximumProfit(vector<int>& profit, vector<int>& weight, int capacity) {
        vector<unordered_map<int, int>> cache(profit.size());
        return dfs(profit, weight, 0, capacity, cache);
    }
    int dfs(const vector<int>& profit, const vector<int>& weight, int index, int capacity, vector<unordered_map<int, int>>& cache) {
        if (index == profit.size() || capacity == 0) {
            return 0;
        }
        if (cache[index].find(capacity) != cache[index].end()) {
            return cache[index][capacity];
        }
        int ifNotPicked = dfs(profit, weight, index + 1, capacity, cache);
        int ifPicked = INT_MIN;

        if (weight[index] <= capacity) {
            ifPicked = profit[index] + dfs(profit, weight, index, capacity - weight[index], cache);
        }

        cache[index][capacity] = max(ifNotPicked, ifPicked);

        return cache[index][capacity];
    }

};

File: Data Structures & Algorithms/coin-change-ii/submission-2.java
class Solution {
    public int change(int amount, int[] coins) {
        Map<String, Integer> dp = new HashMap<>();
        return dfs(coins, amount, 0, dp);
    }

    private int dfs(int[] coins, int amount, int index, Map<String, Integer> dp) {
        if (amount == 0) {
            return 1;
        }
        if (index == coins.length) {
            return 0;
        }

        String key = amount + "," + index;
        if (dp.containsKey(key)) {
            return dp.get(key);
        }

        int ifPicked = 0, ifNotPicked = dfs(coins, amount, index + 1, dp);

        if (coins[index] <= amount) {
            ifPicked = dfs(coins, amount - coins[index], index, dp);
        }

        dp.put(key, (ifPicked + ifNotPicked));
        return dp.get(key);
    }
}
